lib/src/Base/Stat/NonStationaryCovarianceModelFactory.cxx

```diff
@@ -75,100 +75,30 @@ CovarianceModel NonStationaryCovarianceModelFactory::build(const ProcessSample &
-        {
-          Scalar coef = 0.0;
-          for (UnsignedInteger sampleIndex = 0; sampleIndex < size_; ++sampleIndex)
-          {
-            coef += input_[sampleIndex](i, k) * input_[sampleIndex](j, l);
-          } // sampleIndex
-          matrix(k, l) = coef * alpha_;
-        } // l
-      } // k
-    } // index
-  }
-
-  inline void evaluateNonCentered( const TBB::BlockedRange<UnsignedInteger> & r ) const
-  {
-    for (UnsignedInteger index = r.begin(); index != r.end(); ++index)
-    {
-      const UnsignedInteger i = static_cast< UnsignedInteger >(sqrt(2 * index + 0.25) - 0.5);
-      const UnsignedInteger j = index - (i * (i + 1)) / 2;
-      const Point muI(mean_.getValueAtIndex(i));
-      const Point muJ(mean_.getValueAtIndex(j));
-      CovarianceMatrix & matrix(output_[index]);
-      for (UnsignedInteger k = 0; k < dimension_; ++k)
-      {
-        const Scalar muIK = muI[k];
-        for (UnsignedInteger l = 0; l <= k; ++l)
-        {
-          const Scalar muJL = muJ[l];
-          Scalar coef = 0.0;
-          for (UnsignedInteger sampleIndex = 0; sampleIndex < size_; ++sampleIndex)
-          {
-            coef += (input_[sampleIndex](i, k) - muIK) * (input_[sampleIndex](j, l) - muJL);
-          } // sampleIndex
-          matrix(k, l) = coef * alpha_;
-        } // l
-      } // k
-    } // index
-  }
-}; /* end struct ComputeCovariancePolicy */
-
 UserDefinedCovarianceModel NonStationaryCovarianceModelFactory::buildAsUserDefinedCovarianceModel(const ProcessSample & sample, const Bool isCentered) const
 {
   const Mesh & mesh(sample.getMesh());
   const UnsignedInteger N = mesh.getVerticesNumber();
+  // Special case for a sample of size 1
+  if (N == 0) return UserDefinedCovarianceModel(mesh, CovarianceMatrixCollection(0));
   // Create a collection of null CovarianceMatrix
+  // Estimate the covariance function discretization
+  const CovarianceMatrix sigma(buildAsCovarianceMatrix(sample, isCentered));
   const UnsignedInteger size = (N * (N + 1)) / 2;
   const UnsignedInteger dimension = sample.getDimension();
   CovarianceMatrixCollection collection(size);
-  for (UnsignedInteger i = 0; i < size; ++i) collection[i] = CovarianceMatrix(SquareMatrix(dimension).getImplementation());
-
-  // Special case for a sample of size 1
-  if (size == 0) return UserDefinedCovarianceModel(mesh, collection);
-
-  const ComputeCovariancePolicy policy( sample, isCentered, collection );
-  TBB::ParallelFor( 0, size, policy );
+  CovarianceMatrix currentCovariance(dimension);
+  for (UnsignedInteger index = 0; index < size; ++index)
+    {
+      const UnsignedInteger i = static_cast< UnsignedInteger >(sqrt(2 * index + 0.25) - 0.5);
+      const UnsignedInteger j = index - (i * (i + 1)) / 2;
+      for (UnsignedInteger k = 0; k < dimension; ++k)
+	{
+	  std::copy(&sigma(i * dimension, (j + k) * dimension), &sigma(i * dimension, j * dimension) + dimension, &currentCovariance(0, k));
```

Comment:
2nd argument is wrong; this error had been fixed in later commits, thus it would be better to squash commits together.